File: packages/platools-js/src/core/registry.ts
/**
 * Tool registry — tracks every `platools.tool()` registered tool.
 *
 * Mirrors `platools/core/registry.py`. The registry is a plain map
 * behind a thin API so the decorator, transport, and doctor can all
 * share a single view of the consumer's registered tools.
 */

import type { ToolDef } from "../types.js";

export class ToolRegistry {
  private readonly tools: Map<string, ToolDef> = new Map();

  public register(tool: ToolDef): void {
    if (this.tools.has(tool.name)) {
      throw new Error(
        `tool "${tool.name}" is already registered — tool names must be unique`,
      );
    }
    this.tools.set(tool.name, tool);
  }

  public get(name: string): ToolDef | undefined {
    return this.tools.get(name);
  }

  public all(): ToolDef[] {
    return Array.from(this.tools.values());
  }

  public names(): string[] {
    return Array.from(this.tools.keys());
  }

  public has(name: string): boolean {
    return this.tools.has(name);
  }

  public get size(): number {
    return this.tools.size;
  }
}

File: packages/platools-py/platools/core/registry.py
"""Tool registry — tracks every `@platools.tool()` decorated function."""

from __future__ import annotations

from platools.types import ToolDef


class ToolRegistry:
    def __init__(self) -> None:
        self._tools: dict[str, ToolDef] = {}

    def register(self, tool: ToolDef) -> None:
        if tool.name in self._tools:
            raise ValueError(
                f"tool {tool.name!r} is already registered — tool names must be unique"
            )
        self._tools[tool.name] = tool

    def get(self, name: str) -> ToolDef | None:
        return self._tools.get(name)

    def remove(self, name: str) -> bool:
        """Remove a tool from the next complete declaration."""
        return self._tools.pop(name, None) is not None

    def clear(self) -> None:
        """Remove all tools; the next declaration prunes all server mappings."""
        self._tools.clear()

    def all(self) -> list[ToolDef]:
        return list(self._tools.values())

    def names(self) -> list[str]:
        return list(self._tools.keys())

    def __len__(self) -> int:
        return len(self._tools)

    def __contains__(self, name: object) -> bool:
        return name in self._tools
